causal-analysis/TowardsCausalThinking.ipynb

```diff
@@ -0,0 +1,2117 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f67c758e",
+   "metadata": {},
+   "source": [
+    "##### Store Gateway\n",
+    "\n",
+    "Reference: [Existing Thanos Alerts](https://github.com/thanos-io/thanos/blob/main/mixin/alerts/store.libsonnet)\n",
+    "\n",
+    "- ACM does not have all of these alerts today.\n",
+    "\n",
+    "\n",
+    "|Number|Alert|Meaning|\n",
+    "|---:|:---|:---|\n",
+    "|1|ThanosStoreGrpcErrorRate|Thanos Store is failing to handle gRPC requests.|\n",
+    "|2|ThanosStoreSeriesGateLatencyHigh|Thanos Store has high latency for store series gate requests.|\n",
+    "|3|ThanosStoreBucketHighOperationFailures|Thanos Store Bucket is failing to execute operations.|\n",
+    "|4|ThanosStoreObjstoreOperationLatencyHigh|Thanos Store is having high latency for bucket operations.|\n",
+    "\n",
+    "\n",
+    "Notice - all of these are excellent symptoms. It does not explain why these happened.\n",
+    "\n",
+    "__storegw-health-indicator is a vector__\n",
+    "\n",
+    "$\\mathbf{storegw-health-indicator} = \\begin{bmatrix} ThanosStoreGrpcErrorRate \\\\\\ ThanosStoreSeriesGateLatencyHigh \\\\\\ ThanosStoreBucketHighOperationFailures \\\\\\ ThanosStoreObjstoreOperationLatencyHigh   \\end{bmatrix}$\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b9b4934",
+   "metadata": {},
+   "source": [
+    "#### Analysis with Causal Model (aka metrics)\n",
+    "\n",
+    "In this case, we will assign metrics to each of the nodes - and will model relationships between the metrics as mentioned earlier. Based on this, the above causal graphs can be slightly change."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "597034c6",
+   "metadata": {},
+   "source": [
+    "##### Compactor\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "<!--\n",
+    "$compactor-health = \\mathbf{f}(is-compactor-out-ofspace,is-there-a-bad-block)$\n",
+    "-->\n",
+    "$\\mathbf{compactor-health} = \\mathbf{f}(is-compactor-out-ofspace,is-there-a-bad-block)$\n",
+    "\n",
+    "How do we capture if there is a bad block - this may be hard and we may require some approximations perhaps?\n",
+    "A bad compactor is a train wreck waiting to happen. This can quickly snowball into lot more issues if not taken care off right away.\n",
+    "  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "18e19d13",
+   "metadata": {},
+   "source": [
+    "##### Receiver\n",
+    "\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "$\\mathbf{receiver-health} = \\mathbf{f}(time-series, number-of-receiver-replicas, is-receiver-pv-out-of-space, can-receiver-reach-obj-store)$\n"
```

Comment:
Health is mostly determined by active time series https://grafana.com/docs/tempo/latest/metrics-generator/active-series/. Also this is a component with read capability, so it can serve StoreAPI requests from Querier. And Receive can be configured to retain blocks for set amount of time, after which it ships off to objstore. That is also an indicator of health as longer retention periods mean serving heavy queries alongside ingest. Would be good to factor these in as well.
```suggestion
    "$\\mathbf{receiver-health} = \\mathbf{f}(active-time-series, number-of-receiver-replicas, is-receiver-pv-out-of-space, number-of-storeapi-calls, can-receiver-reach-obj-store, retention)$\n"
```